### PR TITLE
Pytest for unit testing

### LINKED ISSUE #3617: Use `pytest` for testing
### Description

I recently learned that `pytest` supports `unittest`-style tests out of the box and needs little to no extra configuration for setting up test commands and just a tad bit more configuration to use without any differences to how the tests have been written.

While rewriting the bulk of our tests in the `tests/` folder and subdirectories to `pytest` ("Migrate to `pytest`", as it would be called) is a humongous task rife with tedium and should be reserved for potential long-term projects, I would say that we are definitely in a position to start using `pytest` itself.

I ran the unit tests on my macOS machine with eight cores locally with `pytest-xdist` installed and enabled, using the command

```python
pytest tests/unit/
```

to run them in parallel, and it offered a staggering `2.5x` speedup over `python run-tests.py --unit`, i.e., serial execution of tests!

### Motivation

1. `pytest` is a modern testing framework used by most packages in the Scientific Python ecosystem
2. We are already using it with `nbmake` to test the example notebooks
3. It simplifies testing a bit since we can let go of most of the code in the `run-tests.py` file and use the `pytest` built-in logging and path discovery functionalities to extract the most out of our testing
4. Better CI times by a very big margin – unit tests and integration tests can both be completed in under ~10 minutes on GitHub Actions runners in separate jobs, which means we can add them to a single job one after the other as well (it reduces the strain on the runners and we wouldn't need to request so many runners like we do currently).
5. Integration tests can be executed to test the built wheels for a Python version with `cibuildwheel`
6. ...and so on

### Possible Implementation

- The `conftest.py` file can be used to configure the `unittest.TestCase` metaclass we currently use for ensuring the reliability of the test cases (see #2833)
- From @Saransh-cpp: `pytest.importorskip` can be used to test the optional dependencies
- I noticed that some Jax solver test cases in the integration tests are causing some troubles when running in parallel mode, we can either reduce the distribution of the workers to run all tests in the same class in serial, or the better way mark just those tests with a certain decorator or `@pytest.fixture` that instructs them to be assigned to a single worker.
- Other nitty-gritties and niceties such as `pytest-cov` and `pytest-doctest`/`pytest-doctestplus` or other plugins out of the hundreds of plugins can be looked into later as a part of the rewriting process

### Additional context

N/A

### PR DESCRIPTION
**Pull request by prady0t**
# Description

Still a WIP. 
Since most of the integration tests were failing, they need to be rewritten completely before we can switch to pytest. For now we are using pytest just for unit tests. We still reduce testing time while running `nox` by a considerable factor.

Addresses a part of #3617 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by prady0t** [comment]
Reason why tests are failing is because `pytest` recognises `arg`, `arg1`, `arg2` from across different function's arguments as [pytest fixtures](https://pytest-with-eric.com/fixtures/pytest-fixtures/#Fixtures-In-The-Test). 
I tried using lambda function in place of 
```py
def test_function(arg):
     return arg + arg
```
using:
```
expr = pybamm.Function(lambda arg: arg+arg, a)
```
insted of 
https://github.com/pybamm-team/PyBaMM/blob/8056d224e75b3926743319a4743abb7c6ab0dac2/tests/unit/test_expression_tree/test_operations/test_evaluate_python.py#L96

But assertion fails (in unittest and pytest):
```
AssertionError: 'function (<lambda>)' != 'function (test_function)'
- function (<lambda>)
+ function (test_function)
```
Maybe if we define `arg fixture` we could solve this.  
@agriyakhetarpal what do you think?

**Comment by prady0t** [comment]
One easy fix is to rename functions like `test_function` to maybe `function1` and others respectively. Due to pytest naming convention, it recognises these as test functions.

**Comment by agriyakhetarpal** [comment]
Does renaming them from `test_function` to something like `function_for_testing` help? Using a lambda function is not only a bit hard on readability but also it would not be able to cover every use case (returns a single expression and you might need to define a different one everywhere)

**Comment by prady0t** [comment]
Why are the CI tasks failing?

**Comment by agriyakhetarpal** [comment]
> Why are the CI tasks failing?

It's probably coming from Matplotlib opening too many figures/plots when running the tests. We should close them after they are opened, I guess. There might be a few other failures due to differences in how `unittest` and `pytest` handle test cases, so you might have to fix the tests.

Also, according to https://github.com/pybamm-team/PyBaMM/pull/3857#issuecomment-2001808198 that the coverage is 19.40%, which means that the coverage percentage is not being reported correctly. We would need to add pytest-cov and configure it with the unit tests, as noted by @Saransh-cpp in https://github.com/pybamm-team/PyBaMM/issues/3617#issuecomment-1854702640.

**Comment by prady0t** [comment]
These are mostly the failures we have to deal with :

```
FAILED tests/unit/test_expression_tree/test_operations/test_latexify.py::TestLatexify::test_latexify -   File "/Users/runner/work/PyBaMM/PyBaMM/pybamm/expression_tree/operations/latexify.py", line 82
    geo_latex = f"\quad {rng_min} < {name} < {rng_max}"
                                                      ^
SyntaxError: invalid escape sequence '\q'
```


```
FAILED tests/unit/test_expression_tree/test_operations/test_latexify.py::TestLatexify::test_latexify_other_variables -   File "/Users/runner/work/PyBaMM/PyBaMM/pybamm/expression_tree/operations/latexify.py", line 82
    geo_latex = f"\quad {rng_min} < {name} < {rng_max}"
                                                      ^
SyntaxError: invalid escape sequence '\q'
```


```
FAILED tests/unit/test_expression_tree/test_operations/test_latexify.py::TestLatexify::test_sympy_preview -   File "/home/runner/work/PyBaMM/PyBaMM/pybamm/expression_tree/operations/latexify.py", line 82
    geo_latex = f"\quad {rng_min} < {name} < {rng_max}"
                                                      ^
SyntaxError: invalid escape sequence '\q'
```


```
FAILED tests/unit/test_util.py::TestUtil::test_import_optional_dependency - AssertionError: "Optional dependency anytree is not available." does not match "Optional dependency anytree.exporter is not available. See https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html#optional-dependencies for more details."
```


```
FAILED tests/unit/test_simulation.py::TestSimulation::test_create_gif - _tkinter.TclError: Can't find a usable init.tcl in the following directories: 
    {C:\hostedtoolcache\windows\Python\3.11.8\x64\tcl\tcl8.6}

C:/hostedtoolcache/windows/Python/3.11.8/x64/tcl/tcl8.6/init.tcl: couldn't read file "C:/hostedtoolcache/windows/Python/3.11.8/x64/tcl/tcl8.6/init.tcl": No error
couldn't read file "C:/hostedtoolcache/windows/Python/3.11.8/x64/tcl/tcl8.6/init.tcl": No error
    while executing
"source C:/hostedtoolcache/windows/Python/3.11.8/x64/tcl/tcl8.6/init.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel #0 [list source $tclfile]"


This probably means that Tcl wasn't installed properly.
```


```
FAILED tests/unit/test_plotting/test_plot.py::TestPlot::test_plot
```

I don't understand  failures with `tcl` and `anytree`.  They fail randomly on different envs.

**Comment by agriyakhetarpal** [comment]
I think the `anytree` error in `import_optional_dependency` is related to an incorrect error message for the dependency (use use `anytree.exporter` specifically, but we import `anytree`). Could you try that test with changes from #3892 applied? It should get merged soon.

Re: the Matplotlib error, I'm not sure either. It looks like an unrelated error and is probably coming from `actions/setup-python`? Does the test work for you locally?

**Comment by prady0t** [comment]
I did encounter `anytree` but not the `test_plot` error. So that might be the case.
Also we can remove `invalid escape sequence` failure by using raw string, instead of:

https://github.com/pybamm-team/PyBaMM/blob/39cff38cc7c424285761101c3e046e0d04af5dff/pybamm/expression_tree/operations/latexify.py#L82

we can do:
```py
        geo_latex = fr"\quad {rng_min} < {name} < {rng_max}"
```
Can we do that without any breaking changes?

**Comment by prady0t** [comment]
We should have more number of checks with this.

**Comment by prady0t** [comment]
How to get covergae report?
I tried running Running 
```
 python -m pytest --cov tests/unit  
``` 
but it's unable to find tests.
Also running 
```
coverage run -m unittest discover tests/unit report  
```
only gives me 20% coverage. What am I doing wrong?

**Comment by Saransh-cpp** [comment]
The command should be:
```
pytest --cov=pybamm tests/unit
```
More on coverage with pytest here - https://learn.scientific-python.org/development/guides/coverage/

**Comment by prady0t** [comment]
> The command should be:
> 
> ```
> pytest --cov=pybamm tests/unit
> ```
> 
> More on coverage with pytest here - https://learn.scientific-python.org/development/guides/coverage/

Oh wrong command. This was embarrassing. 🙃 
I'm now at 96% coverage.

**Comment by prady0t** [comment]
Is the doctests failure related to changes in this PR? Do I need to integrate `xdoctest` for it to pass?

**Comment by kratman** [comment]
> Is the doctests failure related to changes in this PR? Do I need to integrate `xdoctest` for it to pass?

The tests appear to be passing on other branches, so most likely it is something in your PR. If you really get stuck then I can try to help out

**Comment by prady0t** [comment]
> > Is the doctests failure related to changes in this PR? Do I need to integrate `xdoctest` for it to pass?
> 
> The tests appear to be passing on other branches, so most likely it is something in your PR. If you really get stuck then I can try to help out

Let me try integrating pytest plugins for doctest and coverage. This might solve it.

**Comment by agriyakhetarpal** [comment]
The `import pytest` error should get fixed if you add the `[dev]` set of optional dependencies to be installed in the `doctests` and the `docs` `nox` sessions.

**Comment by prady0t** [comment]
Reason for failing tests in py3.12 https://www.python.org/downloads/release/python-3122/#:~:text=Invalid%20backslash%20escape,in%20the%20future.)

**Comment by prady0t** [comment]
Most of the tests are passing now. 😄

**Comment by agriyakhetarpal** [comment]
[This JAX test](https://github.com/pybamm-team/PyBaMM/actions/runs/8390154439/job/22978084660?pr=3857#step:11:3814) has previously failed for us sometimes and looks to be flaky – in a separate PR later on, we can explore some of the [available plugins](https://docs.pytest.org/en/stable/explanation/flaky.html#plugins) and decorate those with `@pytest.mark.flaky()`. For now, I re-triggered the failing job

**Comment by prady0t** [comment]
> [This JAX test](https://github.com/pybamm-team/PyBaMM/actions/runs/8390154439/job/22978084660?pr=3857#step:11:3814) has previously failed for us sometimes and looks to be flaky – in a separate PR later on, we can explore some of the [available plugins](https://docs.pytest.org/en/stable/explanation/flaky.html#plugins) and decorate those with `@pytest.mark.flaky()`. For now, I re-triggered the failing job

Should I open an issue regarding this?

**Comment by agriyakhetarpal** [comment]
The parallel testing looks great – GitHub Actions runners do not have a lot of resources so I guess running locally will reveal the speedups more prominently. I can review this in a while – for now, we will also require updates to the Contributing guide and source installation instructions wherever applicable (while mentioning that the unit tests use `pytest` and the integration tests still use `unittest`)

P.S. the failing Lychee job looks fixable, could you replace it with a different link, i.e., CasADi 2.0.0 isn't really relevant anymore and we could add links to the stable documentation?

**Comment by agriyakhetarpal** [comment]
> Should I open an issue regarding this?

Sure! It would be good to track that

**Comment by prady0t** [comment]
> The parallel testing looks great – GitHub Actions runners do not have a lot of resources so I guess running locally will reveal the speedups more prominently. I can review this in a while – for now, we will also require updates to the Contributing guide and source installation instructions wherever applicable (while mentioning that the unit tests use `pytest` and the integration tests still use `unittest`)
> 
> P.S. the failing Lychee job looks fixable, could you replace it with a different link, i.e., CasADi 2.0.0 isn't really relevant anymore and we could add links to the stable documentation?

Let me look into it. Also is there a reason why running tests on macOS based runners are generally slower? (for me running on macos, py 3.10 locally takes about 12-14 min which is still faster than the time taking by runners but is significantly lot when compared to linux and windows )

**Comment by agriyakhetarpal** [comment]
> Also is there a reason why running tests on macOS based runners are generally slower? (for me running on macos, py 3.10 locally takes about 12-14 min which is still faster than the time taking by runners but is significantly lot when compared to linux and windows )

I ~don't~ _do_ think I'm able to reproduce that – up to 94% completion or so on macOS locally the test suite is quite fast, but then I face some heavy slowdowns when running the tests for the IDAKLU solver, IDAKLUJax tests, and the JAX BDF solver tests – and then with some test cases in `tests/unit/test_meshes/test_one_dimensional_submesh.py` and the `scikits.odes` solvers. But since these tests work fine in the serial implementation, could you try this solution in the meantime: https://github.com/pytest-dev/pytest-xdist/issues/385#issuecomment-1304877301? This is the same thread as the one I had shared earlier on Slack in the `infrastructure` channel when we were discussing `pytest-profiling`. I did see later on that JAX itself is using `pytest-xdist`, so it likely has something to do with our solver rather than being an upstream bug.

Also, it looks like the macOS Python 3.12 tests [are failing](https://github.com/pybamm-team/PyBaMM/actions/runs/8390154439/job/22984682748?pr=3857) for a different reason now, oops! :) Could you try fixing that?

**Comment by kratman** [comment]
> okay, it passes now, but we have to ensure that the order of the tests is deterministic (i.e., `test_util.py::test_import_optional_dependency` and related tests should always be tested at the last).

The tests are parallel by default in PyTest, shouldn't run order be irrelevant? I would expect that PyTest is sort of sandboxing the tests.

If the optional dependencies tests are a problem, then they should not be run with the regular tests. That is more of an integration test than a unit test anyway.

**Comment by kratman** [comment]
I am going to merge develop into this to see if the test failures happen again

File: run-tests.py
Comment (kratman) [inline run-tests.py:15]: How much extra work would it be to remove unittest so it becomes pure pytest?

File: tests/unit/test_expression_tree/test_operations/test_evaluate_python.py
Comment (kratman) [inline tests/unit/test_expression_tree/test_operations/test_evaluate_python.py:288]: Can these be made into raw strings instead of manual escapes or does that break the regex?

File: tests/unit/test_expression_tree/test_operations/test_jac.py
Comment (kratman) [inline tests/unit/test_expression_tree/test_operations/test_jac.py:13]: Seems like these functions repeat a lot, can we move them to a utilities file instead of just renaming them all. Good to clean up a bit when stuff like this is found

**Review by kratman — CHANGES_REQUESTED** [review]
Minor stuff

File: run-tests.py
Comment (kratman) [reply run-tests.py:15]: It can be done in another ticket, I am just curious since right now we are using both libraries for testing

Comment (agriyakhetarpal) [reply run-tests.py:15]: I guess I can answer this a bit – @prady0t had previously mentioned that at least half of the integration tests fail currently with `pytest` – so an estimated timeline could be one chunk of the duration of a standard twelve-week GSoC project (plus or minus a few weeks for any associated technical or non-technical reasons). The tests will have to be rewritten with new syntax either manually or through an AST-based parsing tool.

Comment (kratman) [reply run-tests.py:15]: 12 weeks sounds like a lot, why is it so high?

Comment (agriyakhetarpal) [reply run-tests.py:15]: We would need some other things from the project too. Some of them, off the top of my mind:

1.  Rewriting all the tests (unit too, besides integration) to pytest-style tests
2. Testing the built wheels (`pybamm.test()` similar to what is offered by other projects)
4. Improving test coverage with `hypothesis` (our code coverage is already high, but we could deal with a lot less verbosity when writing tests and check for any uncovered cases)
5. Adding pytest plugins based on ideas from other repositories

so all that does end up being a 175 hour project – and also, a potential mentees are mostly newcomers to open source software projects, so we need to give them time to explore the codebase and learn, so all that adds up

Comment (agriyakhetarpal) [reply run-tests.py:15]: We would need some other things from the project too. Some of them, off the top of my mind:

1.  Rewriting all the tests (unit too, besides integration) to pytest-style tests
2. Testing the built wheels (`pybamm.test()` similar to what is offered by other projects)
4. Improving test coverage with `hypothesis` (our code coverage is already high, but we could deal with a lot less verbosity when writing tests and check for any uncovered cases)
5. Adding pytest plugins based on ideas from other repositories

so all that does end up being a 175 hour project – and also, a potential mentees are mostly newcomers to open source software projects :), so we need to give them time to explore the codebase and learn, so all that adds up

**Comment by agriyakhetarpal** [comment]
To fix the failing Windows tests, let's use the Agg backend for Matplotlib to not open up a GUI. We can either the `MPLBACKEND` environment variable to the `PYBAMM_ENV` dictionary in  `noxfile.py`, or we can do

```python
from matplotlib import use
use("Agg")
```

wherever we have `matplotlib` being imported in the files of the `tests/` folder. The latter would be preferable because it would available without `nox` (i.e., with just `run-tests.py` invocations too).

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:253]: We should not ignore `DeprecationWarning`s, IMO – they serve as reminders for imminent API changes from dependencies upstream. We do not want our CI to break all of a sudden

File: pyproject.toml
Comment (agriyakhetarpal) [inline pyproject.toml:255]: We probably shouldn't ignore `UserWarning`s and `RuntimeWarning`s either, does not ignoring them fail the tests?

File: tests/unit/test_expression_tree/test_operations/test_evaluate_python.py
Comment (prady0t) [reply tests/unit/test_expression_tree/test_operations/test_evaluate_python.py:288]: Yes we can convert all of them to raw string without any breaking change.

File: tests/unit/test_expression_tree/test_operations/test_jac.py
Comment (prady0t) [reply tests/unit/test_expression_tree/test_operations/test_jac.py:13]: in `util.py` file right?

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml:255]: Oh, thanks the clarifications! We can revisit this sometime soon, then.

File: tests/unit/test_expression_tree/test_operations/test_jac.py
Comment (kratman) [reply tests/unit/test_expression_tree/test_operations/test_jac.py:13]: Yeah in a testing utils file, not a main pybamm utils file. Cutting down on duplicated code is just a good thing, especially when you already had to change the names of them all

File: tests/unit/test_expression_tree/test_operations/test_evaluate_python.py
Comment (kratman) [reply tests/unit/test_expression_tree/test_operations/test_evaluate_python.py:288]: Ok then I would go ahead and do it the raw string method

File: tests/unit/test_expression_tree/test_operations/test_jac.py
Comment (agriyakhetarpal) [reply tests/unit/test_expression_tree/test_operations/test_jac.py:13]: For reference, `tests/shared.py` is what we've been using for this purpose

Comment (prady0t) [reply tests/unit/test_expression_tree/test_operations/test_jac.py]: Done!

File: pyproject.toml
Comment (agriyakhetarpal) [reply pyproject.toml:253]: Unresolving this conversation, @prady0t – if we don't have a lot of `DeprecationWarning`s, could you fix them here? If there are many, then feel free to take that up in a follow-up PR.

**Review by agriyakhetarpal — APPROVED** [review]
This looks quite good on my end, and I have just a few additional comments. Let's merge after either @kratman or @Saransh-cpp can review again.

> I noticed that we set MPLBACKEND to "Template" in run-tests.py, which would have set that backend back again when running the tests. Let's switch that environment variable to Agg as well.

This got lost in the conversation, probably – could you please do this too?

**Review by kratman — APPROVED** [review]
I don't see any major issues here, however @agriyakhetarpal has an open discussion, so lets make sure to resolve that before we merge the changes from @prady0t

File: pyproject.toml
Comment (prady0t) [reply pyproject.toml:253]: There are 976 errors but for only one reason.  

DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html

We need to remove `pkg_resources`

Comment (agriyakhetarpal) [reply pyproject.toml:253]: Ah, right. We can't do that at this time due to `pybtex`

**Comment by agriyakhetarpal** [comment]
Almost all good from my end, just the comment about removing the "Template" Matplotlib backend from `run-tests.py` needs to be addressed and a final CI run to be triggered, should be ready to merge after that.

**Review by arjxn-py — APPROVED** [review]
Nice work @prady0t, everything looks nice to me.

**Pull request merged**
